### ISSUE TITLE
Elixir installer on Windows 10 installs Elixir and IEx as broken

### ISSUE THREAD

**Issue by bmitchell-cfs** [elixir-lang/elixir-windows-setup#15]
### Environment

* Elixir & Erlang/OTP versions (elixir --version):

    * Elixir does not run, but Elixir 1.10.3 was selected to be installed.
![image](https://user-images.githubusercontent.com/64422988/85103254-50807380-b1d4-11ea-9555-a93cab144a98.png)

    * The Erlang version, as reported by the installer is 9.2.
![image](https://user-images.githubusercontent.com/64422988/85103300-655d0700-b1d4-11ea-865b-4fc5e9772653.png)

* Operating system: Windows 10, version 1909, OS build 18363.900

### Current behavior

The Windows installer found [here](https://elixir-lang.org/install.html#windows) does not properly install Elixir. The installer reports no errors, but after installation, both iex.bat and elixir.bat fail by crashing. The application werl.exe fails similarly but then automatically closes without the user being able to inspect the crash.

When running `iex.bat` in Windows Terminal with PowerShell, I get:

```
> iex.bat
2020-06-19 02:16:39 Loading of ~ts failed: ~p

        "c:/Program Files (x86)/Elixir/lib/iex/ebin/Elixir.IEx.CLI.beam"
        badfile
2020-06-19 02:16:39 ~s~n
        "beam/beam_load.c(1862): Error loading module 'Elixir.IEx.CLI':\n  This BEAM file was compiled for a later version of the run-time system than 20.\n  To fix this, please recompile this module with an 20 compiler.\n  (Use of opcode 162; this emulator supports only up to 159.)\n"
2020-06-19 02:16:39 crash_report
    initial_call: {supervisor_bridge,user_sup,['Argument__1']}
    pid: <0.49.0>
    registered_name: []
    error_info: {error,undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,"user_sup.erl"},{line,100}]},{user_sup,init,1,[{file,"user_sup.erl"},{line,49}]},{supervisor_bridge,init,1,[{file,"supervisor_bridge.erl"},{line,80}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [kernel_sup,<0.36.0>]
    message_queue_len: 0
    messages: []
    links: [<0.37.0>]
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 987
    stack_size: 27
    reductions: 223
2020-06-19 02:16:39 supervisor_report
    supervisor: {local,kernel_sup}
    errorContext: start_error
    reason: {undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,"user_sup.erl"},{line,100}]},{user_sup,init,1,[{file,"user_sup.erl"},{line,49}]},{supervisor_bridge,init,1,[{file,"supervisor_bridge.erl"},{line,80}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    offender: [{pid,undefined},{id,user},{mfargs,{user_sup,start,[]}},{restart_type,temporary},{shutdown,2000},{child_type,supervisor}]
2020-06-19 02:16:39 crash_report
    initial_call: {application_master,init,['Argument__1','Argument__2','Argument__3','Argument__4']}
    pid: <0.35.0>
    registered_name: []
    error_info: {exit,{{shutdown,{failed_to_start_child,user,{undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,"user_sup.erl"},{line,100}]},{user_sup,init,1,[{file,"user_sup.erl"},{line,49}]},{supervisor_bridge,init,1,[{file,"supervisor_bridge.erl"},{line,80}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}}},{kernel,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,134}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [<0.34.0>]
    message_queue_len: 1
    messages: [{'EXIT',<0.36.0>,normal}]
    links: [<0.34.0>,<0.33.0>]
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 987
    stack_size: 27
    reductions: 243
2020-06-19 02:16:39 std_info
    application: kernel
    exited: {{shutdown,{failed_to_start_child,user,{undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,"user_sup.erl"},{line,100}]},{user_sup,init,1,[{file,"user_sup.erl"},{line,49}]},{supervisor_bridge,init,1,[{file,"supervisor_bridge.erl"},{line,80}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}}},{kernel,start,[normal,[]]}}
    type: permanent
{"Kernel pid terminated",application_controller,"{application_start_failure,kernel,{{shutdown,{failed_to_start_child,user,{undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,\"user_sup.erl\"},{line,100}]},{user_sup,init,1,[{file,\"user_sup.erl\"},{line,49}]},{supervisor_bridge,init,1,[{file,\"supervisor_bridge.erl\"},{line,80}]},{gen_server,init_it,2,[{file,\"gen_server.erl\"},{line,365}]},{gen_server,init_it,6,[{file,\"gen_server.erl\"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,\"proc_lib.erl\"},{line,247}]}]}}},{kernel,start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,kernel,{{shutdown,{failed_to_start_child,user,{undef,[{'Elixir.IEx.CLI',start,[],[]},{user_sup,start_user,3,[{file,"user_sup.

Crash dump is being written to: erl_crash.dump...done
```

When running `elixir.bat --version` in Windows Terminal with PowerShell, I get:

````
> elixir.bat --version
{"init terminating in do_boot",{undef,[{elixir,start_cli,[],[]},{init,start_em,1,[{file,"init.erl"},{line,1085}]},{init,do_boot,3,[{file,"init.erl"},{line,793}]}]}}
init terminating in do_boot ({undef,[{elixir,start_cli,[],[]},{init,start_em,1,[{_},{_}]},{init,do_boot,3,[{_},{_}]}]})

Crash dump is being written to: erl_crash.dump...done
````

### Expected behavior

The expectation is that the installer should just work. The crash message seems to imply a version mismatch with Erlang, but the installer should properly install both Erlang and Elixir with the correct versions.

**Comment by josevalim** [comment]
Hi!

I understand the desire to force Elixir to point to a specific Erlang version, but given that both Elixir and Erlang are development tools, I would be weary of moving to this direction. For example, Elixir developers use `rebar` and `make` to compile Erlang and native packages and they are separate tools that are going to use the Erlang installed in your PATH. You may come across other tools, such as escripts, that you use on your day-to-day as an Elixir developer that will rely on the Erlang version in PATH and they won't work as expected if each of them use a different Erlang version. This means that forcing Elixir to use a hardcoded Erlang version not from PATH is going to lead to issues down the road.

When it comes to PATH, Windows builds the PATH from multiple configurations. It seems erl9.2 was put in a place with higher preference than Elixir's - this is typically the SYSTEM one. But installing on the SYSTEM one is usually seen as bad practice, so if we do this change, then others may not be pleased.

Similarly, you could be using RabbitMQ and us simply overriding Erlang would break your RabbitMQ installation. I don't see an obvious answer for this: you have two projects that want to use different versions of another project and one of them has to "win". I could argue that RabbitMQ should be the one hardcoding the Erlang installation, because it is a "finished product" rather than a development tool, but they may have their reasons too.

All in all, I understand the annoyance when things do not work out of the box, but I am not sure if there is an obvious answer here. This is not even a Windows specific issue, conflicting installs also happens on Unix systems, perhaps even more frequently.

**Comment by chyndman** [comment]
@bmitchell-cfs,

I'll provide context on the installer's implementation then use this to address your summary of the issues.

In order to determine which Erlang versions are installed, if any, the installer scans the registry for subkeys at `HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Ericsson\Erlang` on 64-bit installs and `SOFTWARE\Ericsson\Erlang` on 32-bit installs. If possible, please report what your system has at these subkeys. 

> 1. The Elixir installer should automatically install a compatible version of Erlang if one is not installed.

The installer pulls the latest Erlang version if no Erlang install exists. It is assumed that the latest version is compatible with at least a few of the most recent Elixir releases. I presume the dominant installer use case is to install the latest version of both E's with neither E already present on the system, so it is unlikely we would add the version compatibility information to the installer.

> 2. The Elixir installer should not allow configuration with an incompatible Erlang version.

See answer to 1.

> 3. The Elixir installer should automatically recognize the latest installed version of Erlang.

Agreed. The reason this isn't occurring is possibly due to a defect in the implementation I described above. Based on your screenshots there may an Erlang upstream change that's causing the version parsing to malfunction.

> 4. The Elixir Installer and/or Elixir tools, should properly honor the requested Erlang version configuration and should not use whatever Erlang is returned first in the path.

As @josevalim mentioned, Elixir follows the precedence set by your `PATH`, and we wouldn't want to introduce other incompatibilities by deviating from this behavior.

Maybe the resolution is to add clarifying text in the installer wizard to at least help the user understand what it's doing with this version information.

**Comment by bmitc** [comment]
Hi @josevalim and @chyndman. Apologies for not getting back to this sooner.

I have mainly been setting Elixir up in WSL rather than using it directly in Windows since there is a fair amount of friction on Windows, but today I needed to use a USB device with Elixir, where USB passthrough is not supported in WSL. So I ended up rediscovering this issue since I'm setting up a new Windows 10 computer.

The issue is with the detection. In my case, Erlang 10.6 was already installed by another tool (in fact, it is the same tool as mentioned above: National Instruments software suite). However, National Instruments _does not_ add Erlang to the path. What's happening is that the Elixir installer is detecting the Erlang 10.6 (likely through the registry as mentioned above), and then if selected, the Elixir installer is what adds it to the path. This is mainly a problem if the version is not compatible with Elixir.

So I manually installed the latest version of Erlang after removing the old version from my path and then re-ran the Elixir installer. This time, the Elixir installer did correctly automatically recognize the latest installed version of Erlang, and I was able to get Elixir installed properly on Windows.

I think a fix that could happen is a detection that the latest Erlang install on the computer is not recent enough, prompting the user to download a later or the latest version of Erlang. Either that or the documentation should simply ask the user to manually install a certain version or later of Erlang.

After installing National Instruments software and running the Elixir installer:
![image](https://user-images.githubusercontent.com/65685447/109246582-22622300-77a8-11eb-8426-6e0eeed0e890.png)

After manually installing the latest Erlang version:
![image](https://user-images.githubusercontent.com/65685447/109246603-2beb8b00-77a8-11eb-878f-447f718a6e71.png)
